src/data_api/events.se

```diff
@@ -1,360 +1,420 @@
+    self.events[eventId].automatedReporterOutcome = outcome
 
-def getReportersPaidSoFar(event):
-    return(self.Events[event].reportersPaidSoFarForEvent)
+def disputeAutomatedReport(eventId):
+    self.migrateEventThroughAllForks(eventId)
+    if (self.isEventFinalized(eventId)):
+        ~invalid()
+    if (self.events[eventId].automatedReporterOutcome == 0):
+        ~invalid()
+    if (self.events[eventId].automatedReporterDisputeBondHolder != 0):
+        ~invalid()
+    self.assertEventIsInAutomatedReportDisputePhase(eventId)
+    # self.controller.lookup('reputationToken').transferFrom(msg.sender, self, self.automatedReporterDisputeBondAmount)
+    self.events[eventId].automatedReporterDisputeBondHolder = msg.sender
+    self.events[eventId].automatedReporterOutcome = self.events[eventId].tentativeOutcome
+    self.events[eventId].tentativeOutcome = 0
+
+def disputeLimitedReporters(eventId):
+    self.migrateEventThroughAllForks(eventId)
+    if (self.isEventFinalized(eventId)):
+        ~invalid()
+    if (self.events[eventId].limitedReportersDisputeBondHolder != 0):
+        ~invalid()
+    self.assertEventIsInDisputePhase()
+    # self.controller.lookup('reputationToken').transferFrom(msg.sender, self, self.limitedReportersDisputeBondAmount)
+    self.events[eventId].limitedReportersDisputeBondHolder = msg.sender
+    self.events[eventId].limitedReportersOutcome = self.events[eventId].tentativeOutcome
+    self.events[eventId].tentativeOutcome = 0
+    self.events[eventId].reportingWindow = self.controller.lookup('reporting').getCurrentReportingWindow() + 1
+
+def disputeAllReporters(eventId):
+    self.migrateEventThroughAllForks(eventId)
+    if (self.isEventFinalized(eventId)):
+        ~invalid()
+    if (self.events[eventId].allReportersDisputeBondHolder != 0):
+        ~invalid()
+    self.assertEventIsInDisputePhase()
+    # self.controller.lookup('reputationToken').transferFrom(msg.sender, self, self.allReportersDisputeBondAmount)
+    self.events[eventId].allReportersDisputeBondHolder = msg.sender
+    self.events[eventId].allReportersOutcome = self.events[eventId].tentativeOutcome
+    self.events[eventId].tentativeOutcome = 0
+    self.events[eventId].reportingWindow = self.controller.lookup('reporting').getCurrentReportingWindow() + 1
+
+# will throw if event is not finalized by the time this method returns
+def finalizeEvent(eventId):
+    if (!self.tryFinalizeEvent(eventId)):
+        ~invalid()
 
-def addReportersPaidSoFar(event):
-    self.controller.assertIsWhitelisted(msg.sender)
-    self.Events[event].reportersPaidSoFarForEvent += 1
+# will return 0 if event is not finalized by the time this method returns
+def tryFinalizeEvent(eventId):
+    # FIXME: handle finalizing forked events
+    # FIXME: when the event is finalized, we need to add its fees to the pool for the ending reporting window and we need to add the count of correct outcome tokens to the pool for the window.  These are necessary for fee collection
```

Comment:
```
    # send fees in market to branch
    marketCurrency = INFO.getCurrency(market)
    marketWallet = INFO.getWallet(market)
    marketBalance = marketCurrency.balanceOf(marketWallet)
    marketShareValue = safeFxpMul(MARKETS.getSharesPurchased(market, 1), MARKETS.getCumulativeScale(market))
    if(marketBalance > marketShareValue and MARKETS.getTradingFee(market) != 0):
        marketWallet.transfer(BRANCHES.getBranchWallet(branch, marketCurrency), marketBalance - marketShareValue)

```

Current code on develop for this to give you an idea.  Main meat is `balance Of Market` - `complete sets outstanding`*`range` gives you fees to distribute